libconsensus/Sealer.h

```diff
@@ -110,6 +112,23 @@ class Sealer : public Worker, public std::enable_shared_from_this<Sealer>
         return m_consensusEngine;
     }
 
+    // set consensusEngineFactory to create corresponding consensus engine
+    void setConsensusEngineFactory(ConsensusEngineFactory::Ptr _consensusEngineFactory)
+    {
+        m_consensusEngineFactory = _consensusEngineFactory;
+    }
+
+    virtual void createConsensusEngine(dev::p2p::P2PInterface::Ptr _service,
+        dev::txpool::TxPoolInterface::Ptr _txPool, BlockChainInterface::Ptr _blockChain,
+        dev::sync::SyncInterface::Ptr _blockSync,
+        dev::blockverifier::BlockVerifierInterface::Ptr _blockVerifier,
+        dev::PROTOCOL_ID const& _protocolId, KeyPair const& _keyPair,
+        h512s const& _sealerList = h512s())
+    {
+        m_consensusEngine = m_consensusEngineFactory->createConsensusEngine(_service, _txPool,

```

Comment:
why the engine is not set from outside?

Reply:
By setting m_consensusEngineFactory, we don't need to modify Ledger when the construction of some consensus-engine has  been changed.